bootstrap/pkg/kfapp/coordinator/coordinator.go

```diff
@@ -241,205 +251,393 @@ func NewKfApp(options map[string]interface{}) (kftypes.KfApp, error) {
+// as needed. The function only overrides the KfDef spec if the option isn't already set.
+// The reason we need this is because in 0.5 different command line options were supplied as arguments
+// to different commands (e.g. init & generate) took different command line options.
+// With 0.6 we want to move to a world in which all the options should be stored in app.yaml.
+// Support for the command line otpions is only provided for backwards compatibility until
+// we remove the options.
+func backfillKfDefFromGenerateOptions(kfdef *kfdefsv2.KfDef, options map[string]interface{}) error {
+	if kfdef.Spec.Platform == kftypes.GCP {
+		if options[string(kftypes.EMAIL)] != nil && options[string(kftypes.EMAIL)].(string) != "" {
+			if kfdef.Spec.Email == "" {
+				kfdef.Spec.Email = options[string(kftypes.EMAIL)].(string)
+			} else {
+				log.Warnf("KfDef.Spec.Email is already set; not overwritting with options value")
+			}
+		}
+
+		if kfdef.Spec.IpName == "" {
+			if options[string(kftypes.IPNAME)] != nil && options[string(kftypes.IPNAME)].(string) != "" {
+				kfdef.Spec.IpName = options[string(kftypes.IPNAME)].(string)
+
+			} else if kfdef.Spec.Platform == kftypes.GCP && kfdef.Name != "" {
+				kfdef.Spec.IpName = kfdef.Name + "-ip"
+			}
+
+			log.Warnf("Defaulting Spec.IpName to %v. This is deprecated; "+
+				"IpName should be explicitly set in app.yaml", kfdef.Spec.IpName)
+		}
+
+		if kfdef.Spec.Hostname == "" {
+
+			if options[string(kftypes.HOSTNAME)] != nil && options[string(kftypes.HOSTNAME)].(string) != "" {
+				kfdef.Spec.Hostname = options[string(kftypes.HOSTNAME)].(string)
+			} else if kfdef.Name != "" && kfdef.Spec.Project != "" && kfdef.Spec.Hostname == "" {
+				kfdef.Spec.Hostname = fmt.Sprintf("%v.endpoints.%v.cloud.goog", kfdef.Name, kfdef.Spec.Project)
+			}
+			log.Warnf("Defaulting Spec.Hostame to %v. This is deprecated; "+
+				"Hostname should be explicitly set in app.yaml", kfdef.Spec.Hostname)
+		}
+
+		if kfdef.Spec.Zone == "" {
+			if options[string(kftypes.ZONE)] != nil && options[string(kftypes.ZONE)].(string) != "" {
+				kfdef.Spec.Zone = options[string(kftypes.ZONE)].(string)
+			} else {
+				kfdef.Spec.Zone = kftypes.DefaultZone
+			}
+			log.Warnf("Defaulting Spec.Zone to %v. This is deprecated; "+
+				"Zone should be explicitly set in app.yaml", kfdef.Spec.Zone)
+		}
+	}
+
 	if options[string(kftypes.MOUNT_LOCAL)] != nil {
 		kfdef.Spec.MountLocal = options[string(kftypes.MOUNT_LOCAL)].(bool)
 	}
-	if options[string(kftypes.DELETE_STORAGE)] != nil && kfdef.Spec.Platform == kftypes.GCP {
-		kfdef.Spec.DeleteStorage = options[string(kftypes.DELETE_STORAGE)].(bool)
+
+	return nil
+}
+
+// LoadKfApp is called from subcommands Apply, Delete, Generate and assumes the existence of an app.yaml
+// file which was created by the Init subcommand. It sets options needed by these subcommands
+//
+// TODO(jlewi): This method is deprecated. It is providing backwards compatibility with existing call sites.
```

Comment:
What do you mean by backwards compatibility?
Why can't we just switch, and let generate/apply calls LoadKfAppCfgFile?

Reply:
I'm trying to minimize the amount of code changes in this PR.
So I didn't want to update all the existing call sites in this PR.
So the point of this PR is to create a new entrypoint and then update existing calls in the following PRs.

Reply:
Oh the other point is that the other LoadKfApp takes options which correspond to command line options. I plan on getting rid of those eventually and everything would be supplied by the config file.

So eventually we would just have LoadKfAppCfgFile. but for the time being we need to continue to support getting information via command line flags.